---
title: "PAMES"
author: "Dario Romagnoli, Matteo Benelli"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{PAMES}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

**PAMES** (*Purity Assessment from MEthylation Sites*) is a tool developed
for estimating purity of tumor samples.
The default process for data obtained with Illumina BeadChip
(450k or 27k) is divided into three steps:

1. computation of AUC (`compute_AUC`)
2. selection of informative CpG __sites__ (`select_informative_sites`/`select_informative_sites_ext`)
3. estimation of purity (`compute_purity`)

For data generated by other technologies (such as Bisulphite Sequencing), the
purity estimation requires one preceding additional step:

- reducing the data to CpG regions (`reduce_to_regions`)

Two `data.frames` are required, one reporting the location of each CpG site
(chromosome, position)
and one reporting the location of each genomic regions (chromosome, start, end, optional_name).
These steps convert beta values from several CpG sites
to a single beta value associated to the genomic regions which CpG sites belong
to (CpG sites located outside regions are discarded).

From now on the analysis proceeds as previously described (notice the different `select` function):

1. computation of AUC (`compute_AUC`)
2. selection of informative CpG __regions__ (`select_informative_regions`/`select_informative_regions_ext`)
3. estimation of purity (`compute_purity`)

A set of pre-computed CpG sites (for Illumina 450k) and CpG regions for 14 cancer
types are available with our [PAMESdata](https://github.com/cgplab/PAMESdata) package.

## Examples

Below are showed examples of the steps required to run different kind of analyses.

### Pre-computed informative CpG sites (BRCA_sites) -- 450k, hg19

    BRCA_purity <- compute_purity(BRCA_tumor, BRCA_sites)

### New set of informative CpG sites -- 27k, hg38

    auc_scores <- compute_AUC(tumor_matrix, control_matrix)
    tumor_sites <- select_informative_sites(tumor_table=tumor_matrix, auc=auc_scores, platform=illumina27k_hg38[3:4])
    tumor_purity <- compute_purity(tumor_matrix, cpg_sites, platform=illumina27k_hg38[3:4])

### New set of informative CpG islands

    bs_toy_reduced <- reduce_to_regions(bs_toy_matrix, bs_toy_sites, cpg_islands)
    bs_tumor_toy_matrix <- bs_toy_reduced[, 1:10]
    bs_control_toy_matrix <- bs_toy_reduced[, 11:20]
    auc <- compute_AUC(bs_tumor_toy_matrix, bs_control_toy_matrix)
    bs_toy_regions <- select_informative_regions(bs_tumor_toy_matrix, auc)
    bs_toy_purity <- compute_purity(bs_tumor_toy_matrix, bs_toy_regions, ref_table=cpg_islands)
