#' Select informative CpG sites
#'
#' This function generates a list of informative CpG sites to be used to estimate
#' the purity of a set of tumor samples.
#'
#' Informative sites are divided into \code{hyper} and \code{hypo} depending on
#' their level of methylation with respect to the average beta-score of normal
#' samples. Both sets are used to compute purity.
#'
#' @param tumor_table A matrix of beta-values (percentage) from tumor samples.
#' @param auc A vector of AUC scores generated by \code{compute_AUC}.
#' @param ref_table A data.frame with two columns (chromosome, genomic_coordinates).
#' @param max_sites Maximum number of sites to retrieve (half hyper-, half
#' hypo-methylated) (default = 20).
#' @param min_distance Avoid selection of CpG sites located within this
#' distance from one another (default = 1e6 bps).
#' @param hyper_range A vector of length 2 with minimum lower and upper values
#' required to select hyper-methylated informative sites.
#' @param hypo_range A vector of length 2 with minimum lower and upper values
#' required to select hypo-methylated informative sites.
#' @param method How to select sites: "even" (half hyper-, half hypo-methylated sites),
#' "top" (highest AUC irregardless of hyper or hypomethylation), "hyper" (hyper-methylated sites only),
#' "hypo" (hypo-methylated, sites only).
#' @param percentiles Vector of length 2: lower and upper percentiles to
#' select sites with beta values outside hypo- and hyper-ranges
#' (default = 0,100; 0th and 100th percentiles,
#' i.e. only min and max beta should be outside of ranges).
#' @param return_info Return also a data.frame of all informative sites (for debug purpose).
#' @return A named list of indexes of informative sites ("hyper-" and "hypo-methylated").
#' @importFrom dplyr "%>%"
#' @export
#' @examples
#' ## WARNING: The following code doesn't retrieve any informative site
#' ## Its purpose is to show how to use the tool
#' auc_data <- compute_AUC(tumor_toy_data, control_toy_data)
#' info_sites <- select_informative_sites(tumor_toy_data, auc_data, illumina27k_hg19[3:4])
select_informative_sites <- function(tumor_table, auc, ref_table,
                                     max_sites = 20, min_distance = 1e6, percentiles = c(0, 100),
                                     hyper_range = c(min = 40, max = 90), hypo_range = c(min = 10, max = 60),
                                     method = c("even", "top", "hyper", "hypo"), return_info=FALSE){

    message(sprintf("[%s] # Select informative sites #", Sys.time()))
    # check parameters

    tumor_table <- as.matrix(tumor_table)
    tumor_table <- round(tumor_table)
    storage.mode(tumor_table) <- "integer"


    assertthat::assert_that(nrow(tumor_table) == length(auc))
    assertthat::assert_that(ncol(ref_table) >= 2)
    assertthat::assert_that(is.character(ref_table[[1]]))
    assertthat::assert_that(is.numeric(ref_table[[2]]))
    assertthat::assert_that(nrow(tumor_table) == nrow(ref_table), msg="Number of rows of tumor_table is not equal to the number of rows of ref_table")

    assertthat::assert_that(is.numeric(max_sites))
    assertthat::assert_that(is.numeric(min_distance), min_distance > 0)

    assertthat::assert_that(is.numeric(hyper_range))
    assertthat::assert_that(is.numeric(hypo_range))
    assertthat::assert_that(is.numeric(percentiles))

    assertthat::assert_that(length(hyper_range) == 2)
    assertthat::assert_that(length(hypo_range) == 2)
    assertthat::assert_that(length(percentiles) == 2)

    assertthat::assert_that(all(dplyr::between(hyper_range, 0, 100)))
    assertthat::assert_that(all(dplyr::between(hypo_range, 0, 100)))
    assertthat::assert_that(all(dplyr::between(percentiles, 0, 100)))

    method <- match.arg(method)
    if (method == "even")
        assertthat::assert_that(max_sites %% 2 == 0, msg="method is set to 'even' but max_sites is not even")

    assertthat::assert_that(is.logical(return_info))

    message(sprintf("- Method: %s", method))
    message(sprintf("- Minimum distance between sites: %g bps", min_distance))
    message(sprintf("- Number of sites to retrieve: %i", max_sites))
    message(sprintf("- Hyper-methylated sites range: %i-%i", hyper_range[1], hyper_range[2]))
    message(sprintf("- Hypo-methylated sites range: %i-%i", hypo_range[1], hypo_range[2]))
    message(sprintf("- Percentiles: %g-%g", percentiles[1], percentiles[2]))

    # minimum and maximum beta per site
    message(sprintf("[%s] Compute min-/max- beta scores...", Sys.time()))
    min_beta <- suppressWarnings(apply(tumor_table, 1, quantile, probs = percentiles[1]/100, na.rm = TRUE))
    max_beta <- suppressWarnings(apply(tumor_table, 1, quantile, probs = percentiles[2]/100, na.rm = TRUE))

    message(sprintf("[%s] Select sites...", Sys.time()))
    diff_meth_sites <- dplyr::tibble(Index = seq_along(auc),
                                     AUC = auc,
                                     Max_beta = max_beta,
                                     Min_beta = min_beta,
                                     Chromosome = ref_table[[1]],
                                     Position = ref_table[[2]])
    diff_meth_sites <- dplyr::mutate(diff_meth_sites,
        Type = dplyr::case_when(AUC > .80 & Min_beta < hyper_range[1] & Max_beta > hyper_range[2] ~ "Hyper",
                                AUC < .20 & Min_beta < hypo_range[1]  & Max_beta > hypo_range[2]  ~ "Hypo",
                                TRUE ~ "No_diff"))
    diff_meth_sites <- dplyr::mutate(diff_meth_sites, AUC = dplyr::if_else(Type == "Hypo", 1-AUC, AUC))
    diff_meth_sites <- dplyr::arrange(diff_meth_sites, -AUC)

    sites_hyper <- dplyr::filter(diff_meth_sites, Type == "Hyper")
    sites_hypo  <- dplyr::filter(diff_meth_sites, Type == "Hypo")
    message(sprintf("* Total hyper-methylated sites = %i", nrow(sites_hyper)))
    message(sprintf("* Total hypo-methylated sites = %i", nrow(sites_hypo)))

    # skip sites too close to previously selected sites
    keep_site_hyper <- logical(nrow(sites_hyper))
    kept_sites <- 0
    i <- 1
    if (nrow(sites_hyper) > 0){
        diff_chr <- as.matrix(adist(sites_hyper$Chromosome)) > 0
        above_distance <- as.matrix(dist(sites_hyper$Position)) >= min_distance

        keep_site_hyper[1] <- TRUE
        kept_sites <- kept_sites + 1
        i <- i + 1
    }
    while (kept_sites < max_sites & i <= nrow(sites_hyper)){
        check_all <- all(diff_chr[i, seq_len(i-1)] | above_distance[i, seq_len(i-1)])
        check_all <- check_all[!is.na(check_all)]
        if (length(check_all) > 0 & isTRUE(all(check_all))) {
            keep_site_hyper[i] <- TRUE
            kept_sites <- kept_sites + 1
        }
        i <- i + 1
    }
    top_hyper_sites <- sites_hyper[keep_site_hyper, ]

    keep_site_hypo <- logical(nrow(sites_hypo))
    kept_sites <- 0
    i <- 1
    if (nrow(sites_hypo) > 0){
        diff_chr <- as.matrix(adist(sites_hypo$Chromosome)) > 0
        above_distance <- as.matrix(dist(sites_hypo$Position)) >= min_distance

        keep_site_hypo[1] <- TRUE
        kept_sites <- kept_sites + 1
        i <- i + 1
    }
    while (kept_sites < max_sites & i <= nrow(sites_hypo)){
        check_all <- all(diff_chr[i, seq_len(i-1)] | above_distance[i, seq_len(i-1)])
        if (isTRUE(check_all)) {
            keep_site_hypo[i] <- TRUE
            kept_sites <- kept_sites + 1
        }
        i <- i + 1
    }
    top_hypo_sites <- sites_hypo[keep_site_hypo, ]

    if (method == "even") {
        sites <- list(hyper = top_hyper_sites %>% dplyr::slice(seq_len(max_sites/2)) %>% dplyr::pull(Index),
                      hypo  = top_hypo_sites %>% dplyr::slice(seq_len(max_sites/2)) %>% dplyr::pull(Index))
    } else if (method == "top") {
        top_sites <- dplyr::bind_rows(top_hyper_sites, top_hypo_sites) %>% dplyr::arrange(-AUC) %>% dplyr::slice(seq_len(max_sites))
        sites <- list(hyper = top_sites %>% dplyr::filter(Type == "Hyper") %>% dplyr::pull(Index),
                      hypo  = top_sites %>% dplyr::filter(Type == "Hypo")  %>% dplyr::pull(Index))
    } else if (method == "hyper") {
        sites <- list(hyper = dplyr::slice(top_hyper_sites, seq_len(max_sites)) %>% dplyr::pull(Index))
    } else if (method == "hypo") {
        sites <- list(hypo = dplyr::slice(top_hypo_sites, seq_len(max_sites)) %>% dplyr::pull(Index))
    }
    names(sites$hyper) <-  names(auc)[sites$hyper]
    names(sites$hypo)  <-  names(auc)[sites$hypo]
    message(sprintf("* Retrieved hyper-methylated sites = %i", length(sites$hyper)))
    message(sprintf("* Retrieved hypo-methylated sites = %i", length(sites$hypo)))

    message(sprintf("[%s] Done", Sys.time()))
    if (return_info) {
        sites_hyper$Keep <- keep_site_hyper
        sites_hypo$Keep <- keep_site_hypo
        return(c(sites, list(info=rbind(sites_hyper, sites_hypo))))
    } else {
        return(sites)
    }
}
